node/pkg/accounting/accounting.go

```diff
@@ -0,0 +1,373 @@
+	msgId := msg.MessageIDString()
+	acct.logger.Info("acct: debug: in SubmitObservation", zap.String("msgID", msgId))
+	// We only care about token bridges.
+	tbk := tokenBridgeKey{emitterChainId: msg.EmitterChain, emitterAddr: msg.EmitterAddress}
+	if _, exists := acct.tokenBridges[tbk]; !exists {
+		if msg.EmitterChain != vaa.ChainIDPythNet {
+			acct.logger.Info("acct: ignoring vaa because it is not a token bridge", zap.String("msgID", msgId))
+		}
+
+		return true, nil
+	}
+
+	// We only care about transfers.
+	if !vaa.IsTransfer(msg.Payload) {
+		acct.logger.Info("acct: ignoring vaa because it is not a transfer", zap.String("msgID", msgId))
+		return true, nil
+	}
+
+	acct.mutex.Lock()
+	defer acct.mutex.Unlock()
+
+	gs := acct.gst.Get()
+	if gs == nil {
+		acct.logger.Error("acct: failed to look up guardian set, blocking publishing", zap.String("msgID", msgId))
+		return false, nil
+	}
+
+	digest := msg.CreateDigest()
+
+	// If this is already pending, don't send it again.
+	if oldEntry, exists := acct.pendingTransfers[msgId]; exists {
+		if oldEntry.digest != digest {
+			digestMismatches.Inc()
+			acct.logger.Error("acct: digest in pending transfer has changed, dropping it",
+				zap.String("msgID", msgId),
+				zap.String("oldDigest", oldEntry.digest),
+				zap.String("newDigest", digest),
+			)
+		} else {
+			acct.logger.Info("acct: blocking transfer because it is already outstanding", zap.String("msgID", msgId))
+		}
+		return false, nil
+	}
+
+	/////////////////////////////////////////////////////////////////////////////
+	// WETH on BSC is 0x3D9E7a12DAa29a8B2b1BFaA9DC97ce018853Ab31
+	// if msg.EmitterChain == vaa.ChainIDBSC {
+	// 	data, err := hex.DecodeString("0000000000000000000000000000000000002386F26FC100002386F26FC10000")
+	// 	if err != nil {
+	// 		acct.logger.Error("acct: debug: failed to decode bogus amount", zap.Error(err))
+	// 	} else {
+	// 		acct.logger.Info("acct: debug: old payload", zap.String("bytes", hex.EncodeToString(msg.Payload)))
+	// 		newPayload := append(msg.Payload[0:1], data...)
+	// 		newPayload = append(newPayload, msg.Payload[33:]...)
+	// 		msg.Payload = newPayload
+	// 		acct.logger.Info("acct: debug: new payload", zap.String("bytes", hex.EncodeToString(msg.Payload)))
+	// 	}
+	// }
+	/////////////////////////////////////////////////////////////////////////////
+
+	// Add it to the pending map and the database.
+	if err := acct.addPendingTransfer(msgId, msg, digest); err != nil {
+		acct.logger.Error("acct: failed to persist pending transfer, blocking publishing", zap.String("msgID", msgId), zap.Error(err))
+		return false, err
+	}
+
+	acct.logger.Info("acct: submitting transfer to accounting for approval", zap.String("msgID", msgId), zap.Bool("canPublish", !acct.enforceFlag))
+
+	// This transaction may take a while. Run it as a go routine so we don't block the processor.
+	if acct.env != GoTestMode {
+		go acct.submitObservationToContract(msg, gs.Index)
```

Comment:
+1, if multiple tx's are submitted in parallel, there will be sequence number collisions.

However, only submitting 1 message per 1 tx per 1 block I think will run into a bottleneck pretty quickly.

I think we'll need to submit a variable number of messages per tx.  So the new goroutine should consume ~100-200 messages at a time from the channel and submit them in a single tx per block.
Multiple observations/messages can be included here in the TX:
https://github.com/wormhole-foundation/wormhole/blob/9fe03b07b27f894d95d4e8cbefb0660046272964/node/pkg/wormconn/send_tx.go#L37

This could make error handling more tricky.  If one of the messages gets an error, I believe the whole tx will be rejected.

Reply:
If we do submit a tx with a 100 messages, we should have a way to recover if 1 of the messages is problematic (causing an error).

I think the ideal way would be to simulate the transaction first, then see if of the messages caused an error.  If so, drop it, then proceed with broadcasting the TX.

Reply:
Nevermind, multiple observations can be included in the message already.

Maybe it'd be useful then that the go routine merges multiple messages into one?  I.e.  If there's 10 messages with 1 observation each, we should submit 1 message with 10 observations instead.